File: problem9/c-solution/p9.c
/*
 * p9.c
 *
 * Zeke Reyna
 *
 * Problem 9:
 * There exists one pythagorean triplet such that a + b + c = 100.
 * Find the product a * b * c.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define		LIMIT		1000
#define		SQR(x)		((x)*(x))

int find_c(int a, int b)
{
	int c = (int) sqrt(SQR(a) + SQR(b));
	if (SQR(c) != SQR(a) + SQR(b))
		c = -1;
	return c;
}

int main(int argc, char **argv)
{
	int sum_limit;
	int i, j, c;
	int product;

	(argc == 2) ? (sum_limit = atoi(argv[1])) : (sum_limit = LIMIT);

	product = -1;
	for (i = 1; i < sum_limit; i++) {
		for (j = 1; j < sum_limit; j++) {
			c = find_c(i, j);
			if (c != -1 && i + j + c == sum_limit)
				goto found;
		}
	}
found:
	product = i * j * c;
	printf("\nThe product of the pythagorean triplets that sum to ");
	printf("%d is %d\n\n", sum_limit, product);
	return 0;
}

File: problem9/python-solution/p9.py
### p9.py
#
# Zeke Reyna
#
# Problem 9:
# There exists one pythagorean triplet such that a + b + c = sum.
# Find the product a * b * c.

import sys
import math

def find_c(a, b):
    c = int(math.sqrt(a*a + b*b))
    if c*c != a*a + b*b:
        c = -1
    return c

if __name__ == "__main__":
    if len(sys.argv) == 2:
        sum = int(sys.argv[1])
    else:
        sum = 1000

    product = -1
    for a in range(1, sum):
        for b in range(1, sum):
            c = find_c(a, b)
            if c != -1 and (a + b + c) == sum:
                product = a * b * c
                break
        if product > 0:
            break

    print("\nThe product of the pythagorean triplets that sum",
          "to", sum, "is", product, "\n")
